Wallet:  Fix Non-Ranged Descriptors with Range [0,0] Trigger Unexpected Wallet Errors in AddWalletDescriptor

Eunovo opened the pull request:
Closes https://github.com/bitcoin/bitcoin/issues/31728

This PR updates the `DescriptorScriptPubKeyMan` to skip range checks for non-ranged descriptors, which previously caused errors when updating a non-ranged descriptor with the range [0,0]

#### Testing
A unit test was added to test the new behaviour

Eunovo committed:
Skip range verification for non-ranged desc

Non-range desc  are always added to the wallet with the range [0,0]. After the descriptor is added, the wallet will  TopUp the keypool. For non-range descriptors, this process updates the desc range to [0,1].

Any attempts to update this non-range descriptor with a [0,0] range will result in an error because the range checks rejects new ranges not included in the old range.

Since  this is a non-range desc, the range information should be disregarded and AddWalletDescriptor should always succeed regardless of provided range information

Eunovo committed:
Test updating non-ranged descriptor with [0,0] range succeeds

w0xlt reviewed (commented):
Concept ACK.

Functional tests might be necessary.

Eunovo commented:
> Functional tests might be necessary.

The `importdescriptors` RPC sets the range for all non-ranged descriptors to `[0,1]` in every request so this error cannot be triggered using the RPC.

In https://github.com/bitcoin/bitcoin/issues/31728 @JeremyRubin explains that this error may be unexpectedly triggered when " experimenting with non-ranged descriptors in custom wallet workflows."

shahsb commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -73,6 +73,27 @@ static void AddKey(CWallet& wallet, const CKey& key)
+            auto descs{Parse(desc_str, provider, error, /* require_checksum=*/ false)};
+            auto& desc{descs.at(0)};
+            WalletDescriptor w_desc{std::move(desc), 0, 0, 0, 0};
+            auto spk_manager{*Assert(wallet.AddWalletDescriptor(w_desc, provider, "", false))};
```
I do not see the call to the function `DescriptorScriptPubKeyMan::CanUpdateToWalletDescriptor()` being made in this test case?

Eunovo replied:
Calling AddWalletDescriptor with a descriptor that's already in wallet, calls UpdateWalletDescriptor which calls CanUpdateWalletDescriptor.  This is what happens on the second call to AddWalletDescriptor.

shahsb commented:
Concept ACK

achow101 commented:
ACK fa0993462b3bc8db40c1a6ea2e7f5fd4a22353a6

BrandonOdiwuor reviewed (commented):
Code Review ACK fa0993462b3bc8db40c1a6ea2e7f5fd4a22353a6

rkrux reviewed (commented):
> The importdescriptors RPC sets the range for all non-ranged descriptors to [0,1] in every request so this error cannot be triggered using the RPC.

I checked that the `importdescriptors` adds a 1 to range end if a range is given in the request.

> that this error may be unexpectedly triggered when " experimenting with non-ranged descriptors in custom wallet workflows.

The issue seems quite peculiar in that it highlights an issue within an internal function instead of the full functionality, makes sense to verify the fix using a unit test.

rkrux commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -73,6 +73,27 @@ static void AddKey(CWallet& wallet, const CKey& key)
+        wallet.SetWalletFlag(WALLET_FLAG_DESCRIPTORS);
+        auto key{GenerateRandomKey()};
+        auto desc_str{"combo(" + EncodeSecret(key) + ")"};
+        for (size_t i = 0; i < 2; i++)
```
Nit: Although the test name is present but it might be helpful to mention why the loop is there, i.e., to update the descriptor by calling `AddWalletDescriptor` multiple times with the same descriptor.

Eunovo replied:
Decided to simplify the test. See https://github.com/bitcoin/bitcoin/pull/32344/commits/97d383af6d54b463da64f45680a146d7e93eb146

rkrux commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2852,6 +2852,11 @@ bool DescriptorScriptPubKeyMan::CanUpdateToWalletDescriptor(const WalletDescript
+    if (!descriptor.descriptor->IsRange()) {
+        // Skip range check for non-range descriptors
+        return true;
+    }
+
     if (descriptor.range_start > m_wallet_descriptor.range_start ||
```
Nit: Tying these 2 if blocks with an `else if` would be nice, although not necessary because the first if returns.

Eunovo replied:
I wasn't satisfied with how this would look, so I left it as-is.

rkrux reviewed (commented):
Concept ACK fa0993462b3bc8db40c1a6ea2e7f5fd4a22353a6

I notice while running the unit tests that it warns about a lack of assertions. Maybe it's expecting a boost check?

```bash
wallet/test/wallet_tests.cpp:76: Entering test case "update_non_range_descriptor"
Test case wallet_tests/update_non_range_descriptor did not check any assertions
wallet/test/wallet_tests.cpp:76: Leaving test case "update_non_range_descriptor"; testing time: 38813us
```

Eunovo commented:
Rebased https://github.com/bitcoin/bitcoin/pull/32344/commits/fa0993462b3bc8db40c1a6ea2e7f5fd4a22353a6 to https://github.com/bitcoin/bitcoin/pull/32344/commits/97d383af6d54b463da64f45680a146d7e93eb146

Simplified update non-range descriptor unit test and fixed this warning:

> I notice while running the unit tests that it warns about a lack of assertions. Maybe it's expecting a boost check?
> 
> ```shell
> wallet/test/wallet_tests.cpp:76: Entering test case "update_non_range_descriptor"
> Test case wallet_tests/update_non_range_descriptor did not check any assertions
> wallet/test/wallet_tests.cpp:76: Leaving test case "update_non_range_descriptor"; testing time: 38813us
> ```

rkrux commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -73,6 +73,25 @@ static void AddKey(CWallet& wallet, const CKey& key)
+        auto descs{Parse(desc_str, provider, error, /* require_checksum=*/ false)};
+        auto& desc{descs.at(0)};
+        WalletDescriptor w_desc{std::move(desc), 0, 0, 0, 0};
+        BOOST_CHECK(wallet.AddWalletDescriptor(w_desc, provider, "", false));
+        // Wallet should update the non-range descriptor successfully
+        BOOST_CHECK(wallet.AddWalletDescriptor(w_desc, provider, "", false));
```
Nice, I did consider removing the loop in the previous review but decided against it so as to simulate a similar flow like this one in a loop: https://github.com/bitcoin/bitcoin/blob/64697529523705f5a411ca893060429afcd5cd47/src/wallet/wallet.cpp#L4131-L4132

However, I did check now that in `GetDescriptorScriptPubKeyMan` the `ScriptPubKeyMan` is gotten via `descriptor.id`, which would remain same whether the `w_desc` reference is same or not.
https://github.com/bitcoin/bitcoin/blob/64697529523705f5a411ca893060429afcd5cd47/src/wallet/wallet.cpp#L3686-L3688

So, overall fine with this approach.

rkrux reviewed (approved):
ACK 97d383a

furszy reviewed (commented):
It would be good to describe the steps to replicate the issue via RPC.
Including a range in a non-ranged descriptor import should cause the process to fail before reaching the wallet descriptor update function.

This step-by-step description would also allow us to convert the unit test into a functional test, which would be ideal, as it's simpler to maintain.

Eunovo commented:
> Including a range in a non-ranged descriptor import should cause the process to fail before reaching the wallet descriptor update function.

AFAICT this will cause `importdescriptor` to throw an error with the message "Range should not be specified for an un-ranged descriptor"

furszy commented:
> > Including a range in a non-ranged descriptor import should cause the process to fail before reaching the wallet descriptor update function.
> 
> AFAICT this will cause `importdescriptor` to throw an error with the message "Range should not be specified for an un-ranged descriptor"

Then how can the user trigger the error being solved here? If it can't be triggered by the user, we're talking about an extra guard rather than a fix — which is fine as well, but it would be better to label the PR differently, since it won't need to be backported.

rkrux commented:
> Then how can the user trigger the error being solved here? If it can't be triggered by the user, we're talking about an extra guard rather than a fix — which is fine as well, but it would be better to label the PR differently, since it won't need to be backported.

Agree that the PR can be labeled differently.

> When attempting to add a non-ranged descriptor with a start/end range of [0,0] via AddWalletDescriptor, an error may appear

The issue #31728 description states that it is triggered via `AddWalletDescriptor`, which is an internal function & not exposed to the end user. I agree with the expectation set in the issue and this corresponding fix does seem to improve the internal functions and their understanding. 

> it can lead to confusion and error messages when using or experimenting with non-ranged descriptors in custom wallet workflows.

I feel that the issue need not labeled as a 'Bug' because the author mentions it is triggered while experimenting with custom flows.

achow101 commented:
ACK 97d383af6d54b463da64f45680a146d7e93eb146
